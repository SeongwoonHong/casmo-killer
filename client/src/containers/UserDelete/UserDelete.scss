.Delete {
    &__form {
        overflow: auto;
        &__btn {
            float: right;
            margin: 15px;
        }
    }
}
