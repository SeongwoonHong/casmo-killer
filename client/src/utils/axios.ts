import axios from 'axios';

axios.interceptors.request.use((config) => {
  console.log(config);
  return config;
});

function setTokenToHeader(header, csrfToken) {
  axios.defaults.headers.common[header] = csrfToken;
}

export {
  axios,
  setTokenToHeader,
};
