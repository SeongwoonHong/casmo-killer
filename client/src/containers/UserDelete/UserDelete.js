import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import axios from 'axios';

import FormMessage from '@sharedComponents/FormMessage';

import UserPageContainer from '../../components/UserPageContainer';
import UserInputField from '../../components/UserInputField';

import './UserDelete.scss';

class Delete extends Component {

  constructor(props) {

    super(props);

    this.state = {
      isLoading: false,
      isSuccess: false,
      agreed: false,
      payload: '',
      message: ''
    };

    this.onChangeHandler = this.onChangeHandler.bind(this);
    this.onSubmitHandler = this.onSubmitHandler.bind(this);

  }

  onChangeHandler(e) {
    this.setState({ payload: e.value });
  }

  async onSubmitHandler() {

    if (!this.state.agreed) {
      this.setState({ agreed: true });
      return;
    }

    this.setState({ isLoading: true });

    if (this.state.payload.length === 0) {

      this.setState({
        isSuccess: false,
        message: 'Please enter your information to continue.'
      });

    } else {

      try {

        const { data } = await axios.delete('/api/user/delete/account', {
          data: {
            payload: this.state.payload
          }
        });

        if (data && data.message) {

          this.setState({
            isSuccess: true,
            agreed: false,
            message: data.message
          });

          this.props.removeUser();

        } else {

          this.setState({
            isSuccess: false,
            message: 'Failed to communicate with the server.'
          });

        }

      } catch (error) {

        console.error(error);
        this.setState({
          isSuccess: false,
          message: error.response && error.response.data.message
        });

      }

    }

    this.setState({ isLoading: false });

  }

  render() {

    const {
      isLoading, isSuccess, agreed, payload, message
    } = this.state;

    const { user, logout } = this.props;

    const button = (isComplete, agree, onClick) => {
      if (isComplete) {
        return (
          <Link
            to="/"
            className="user-form-button"
            onClick={ onClick }>
            Ck-Board Home
          </Link>
        );
      }
      return (
        <button
          type="submit"
          className="user-form-button">
          { agree ? 'Delete My Account' : 'Agree' }
        </button>
      );
    };

    return (
      <UserPageContainer
        className="Delete"
        title="Account Delete"
        icon="delete_forever"
        isLoading={ isLoading }
        formTitle={
          !agreed
            ? 'Confirm Account Deletion'
            : 'Please Confirm Your Information'
        }
        formMsg={
          !agreed
            ? 'This will permanently delete your account. This process is irreversible.'
            : null
        }
        onSubmit={ this.onSubmitHandler }
        button={ button(
          isSuccess && message.length > 0,
          agreed,
          logout)
        }>
        <FormMessage
          message={ message }
          type={ isSuccess ? 'success' : 'error' } />
        <UserInputField
          isVisible={ agreed }
          name={ user.strategy === 'local' ? 'password' : 'email' }
          onChange={ this.onChangeHandler }
          value={ payload } />
      </UserPageContainer>
    );

  }

}

export default Delete;
