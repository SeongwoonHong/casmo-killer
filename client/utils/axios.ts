import axios from 'axios';

const baseURL = 'http://localhost:9000';

axios.defaults.baseURL = baseURL + '/api';

export { axios };
