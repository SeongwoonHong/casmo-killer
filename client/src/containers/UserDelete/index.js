import { connect } from 'react-redux';
import * as actions from '@actions';

import UserDelete from './UserDelete';

function mapStateToProps(state) {
  return {
    user: state.user.user
  };
}

function mapDispatchToProps(dispatch) {
  return {
    removeUser: () => dispatch(actions.removeUser()),
    logout: () => dispatch(actions.logout())
  };
}

export default connect(mapStateToProps, mapDispatchToProps)(UserDelete);
