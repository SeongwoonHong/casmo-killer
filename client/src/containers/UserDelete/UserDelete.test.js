/* eslint react/jsx-boolean-value: 0 */
import React from 'react';
import { shallow } from 'enzyme';
import sinon from 'sinon';
import axios from 'axios';

import UserDelete from './UserDelete';

describe('<UserDelete />', () => {

  let component;
  let mockDelete;

  beforeEach(() => {

    const user = { strategy: 'local' };

    component = shallow(
      <UserDelete user={ user } />
    );

    mockDelete = sinon.stub(axios, 'delete');

  });

  afterEach(() => {
    mockDelete.restore();
  });

  it('should render the component consistently', () => {
    expect(component).toMatchSnapshot();
  });

  it('should set the agreed state to true for further actions', () => {
    component.instance().onSubmitHandler();
    expect(component.state('agreed')).toEqual(true);
  });

  it('should make a DELETE request and resolve the response from the server', async () => {

    const spy = sinon.spy();

    const responseData = { data: { message: 'server messsage' } };
    const promise = Promise.resolve(responseData);

    mockDelete.callsFake(() => promise);

    const wrapper = shallow(
      <UserDelete
        user={ { strategy: 'local' } }
        removeUser={ spy } />
    );

    wrapper.setState({
      agreed: true,
      payload: 'delete payload'
    });
    wrapper.instance().onSubmitHandler();

    await promise;

    expect(wrapper.state('isSuccess')).toEqual(true);
    expect(wrapper.state('agreed')).toEqual(false);
    expect(wrapper.state('message')).toEqual(responseData.data.message);
    expect(spy.calledOnce).toEqual(true);

  });

});
