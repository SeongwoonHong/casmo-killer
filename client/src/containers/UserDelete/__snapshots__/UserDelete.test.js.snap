// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UserDelete /> should render the component consistently 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Delete
    user={
        Object {
            "strategy": "local",
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "button": <button
        className="User__form__btn Delete__form__btn"
        type="submit"
>
        Agree
</button>,
      "children": Array [
        <FormMessage
          className=""
          message=""
          type="error"
/>,
        <UserInputField
          disabled={false}
          isVisible={false}
          message=""
          name="password"
          onChange={[Function]}
          placeholder=""
          title=""
          type=""
          value=""
/>,
      ],
      "className": "Delete",
      "disabled": false,
      "formMsg": "This will permanently deactivate your account. This process is irreversible.",
      "formTitle": "Confirm Account Deactivation",
      "icon": "delete_forever",
      "isLoading": false,
      "onSubmit": [Function],
      "title": "Deactivate Your Account",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "className": "",
          "message": "",
          "type": "error",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": null,
          "disabled": false,
          "isVisible": false,
          "message": "",
          "name": "password",
          "onChange": [Function],
          "placeholder": "",
          "title": "",
          "type": "",
          "value": "",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "button": <button
          className="User__form__btn Delete__form__btn"
          type="submit"
>
          Agree
</button>,
        "children": Array [
          <FormMessage
            className=""
            message=""
            type="error"
/>,
          <UserInputField
            disabled={false}
            isVisible={false}
            message=""
            name="password"
            onChange={[Function]}
            placeholder=""
            title=""
            type=""
            value=""
/>,
        ],
        "className": "Delete",
        "disabled": false,
        "formMsg": "This will permanently deactivate your account. This process is irreversible.",
        "formTitle": "Confirm Account Deactivation",
        "icon": "delete_forever",
        "isLoading": false,
        "onSubmit": [Function],
        "title": "Deactivate Your Account",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "className": "",
            "message": "",
            "type": "error",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": null,
            "disabled": false,
            "isVisible": false,
            "message": "",
            "name": "password",
            "onChange": [Function],
            "placeholder": "",
            "title": "",
            "type": "",
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
